tokio/src/net/windows/named_pipe.rs

```diff
@@ -0,0 +1,1305 @@
+        pub type NamedPipe = crate::doc::NotDefinedHere;
+    }
+}
+
+use self::doc::*;
+
+/// A [Windows named pipe].
+///
+/// Constructed using [NamedPipeClientOptions::create] for clients, or
+/// [NamedPipeOptions::create] for servers. See their corresponding
+/// documentation for examples.
+///
+/// Connecting a client involves a few steps. First we must try to [create], the
+/// error typically indicates one of two things:
+///
+/// * [std::io::ErrorKind::NotFound] - There is no server available.
+/// * [ERROR_PIPE_BUSY] - There is a server available, but it is busy. Sleep for
+///   a while and try again, or use [wait_named_pipe] until it becomes available.
+///
+/// So a typical client connect loop will look like the this:
+///
+/// ```no_run
+/// use std::time::Duration;
+/// use tokio::net::windows::NamedPipeClientOptions;
+/// use tokio::time;
+/// use winapi::shared::winerror;
+///
+/// const PIPE_NAME: &str = r"\\.\pipe\named-pipe-idiomatic-client";
+///
+/// # #[tokio::main] async fn main() -> std::io::Result<()> {
+/// let client = loop {
+///     match NamedPipeClientOptions::new().create(PIPE_NAME) {
+///         Ok(client) => break client,
+///         Err(e) if e.raw_os_error() == Some(winerror::ERROR_PIPE_BUSY as i32) => (),
+///         Err(e) => return Err(e),
+///     }
+///
+///     time::sleep(Duration::from_millis(50)).await;
+/// };
+///
+/// /* use the connected client */
+/// # Ok(()) }
+/// ```
+///
+/// A client will error with [std::io::ErrorKind::NotFound] for most creation
+/// oriented operations like [create] or [wait_named_pipe] unless at least once
+/// server instance is up and running at all time. This means that the typical
+/// listen loop for a server is a bit involved, because we have to ensure that
+/// we never drop a server accidentally while a client might want to connect.
+///
+/// ```no_run
+/// use std::io;
+/// use std::sync::Arc;
+/// use tokio::net::windows::NamedPipeOptions;
+/// use tokio::sync::Notify;
+///
+/// const PIPE_NAME: &str = r"\\.\pipe\named-pipe-idiomatic-server";
+///
+/// # #[tokio::main] async fn main() -> std::io::Result<()> {
+/// // The first server needs to be constructed early so that clients can
+/// // be correctly connected. Otherwise calling .wait will cause the client to
+/// // error.
+/// //
+/// // Here we also make use of `first_pipe_instance`, which will ensure that
+/// // there are no other servers up and running already.
+/// let mut server = NamedPipeOptions::new()
+///     .first_pipe_instance(true)
+///     .create(PIPE_NAME)?;
+///
+/// let shutdown = Arc::new(Notify::new());
+/// let shutdown2 = shutdown.clone();
+///
+/// // Spawn the server loop.
+/// let server = tokio::spawn(async move {
+///     loop {
+///         // Wait for a client to connect.
+///         let connected = tokio::select! {
+///             connected = server.connect() => connected,
+///             _ = shutdown2.notified() => break,
+///         };
```

Comment:
I'm not sure we need to include shutdown here. Calling `abort` on a `JoinHandle` would do it if its just one task, and when we do talk about shutdown, we typically recommend a broadcast channel, as it better handles shutdown of tasks started after the shutdown signal was sent.

Reply:
Removed the shutdown logic completely to reduce clutter.